src/rest_api/classes/variation/widgets/molecular_details.clj

```diff
@@ -0,0 +1,315 @@
+             :sl (seq (:variation/source-location variation))
+             :mutant_header (str name " with " flank " bp flanks")
+             :placeholder nil})
+   :description "wild type and variant sequences in genomic context"})
+
+(defn flanking-pcr-products [variation]
+  {:data nil
+   :description "PCR products that flank the variation"})
+
+;test WBVar01112111
+(defn variation-type [variation]
+  (let [variation-type-map {:variation/engineeered-allele "Engineered allele"
+                            :variation/allele"Allele"
+                            :variation/snp "SNP"
+                            :variation/confirmed-snp "Confirmed SNP"
+                            :variation/predicted-snp "Predicted SNP"
+                            :variation/transposon-insertion "Transposon Insertion"
+                            :variation/natural-variant "Natural Variant"}
+        type-of-mutation-map {:variation/substitution "Substitution"
+                              :variation/insertion "Insertion"
+                              :variation/detetion "Deletion"
+                              :variation/tandem-duplication "Tandem Duplication"}]
+    {:data {:pysical_class (cond
+                             (contains? variation :variation/transposon-insertion)
+                             "Transposon insertion"
+
+                             (contains? variation :variation/transposon-excision)
+                             "Transposon excision"
+
+                             :else
+                             (->> type-of-mutation-map
+                                  (map
+                                    #(str (if
+                                            (contains? variation (first %))
+                                            (second %))))
+                                  (filter #(not= % ""))
+                                  (str/join "/")))
+            :general_class (->> variation-type-map
+                                (map
+                                  #(str (if
+                                          (true? ((first %) variation))
+                                          (second %))))
+                                (filter #(not= % "")))}
+     :description "the general type of the variation"}))
+
+(defn- count-affects [variation]
+  (reduce
+    +
+    (map count [(:variation/gene variation)
+                (:variation/predicted-cds variation)
+                (:variation/transcript variation)
+                (:variation/pseudogene variation)
+                (:variation/feature variation)
+                (:variation/interactor variation)])))
+
+(defn- create-fragment [peptide position]
+  (str (- position 19)
+       "..."
+       (subs peptide (- position 20) 19)
+       " "
+       "<b>" (subs peptide (- position 1) 1) "</b>"
+       " "
+       (subs peptide position 20)
+       "..."
+       (+ position 19)))
+
+(defn- do_markup [peptide var-start variation mutation-type]
```

Comment:
Not sure why this stuff isn't done in client-side js, I guess the perl code did, so you've just re-produced?

Reply:
Yes, especially this molecular details widget turned into a rabbit hole. I think the main reason is because this logic needs to be applied to the data before the perl template works on it. If we were to start using React js for rendering the widgets we would be able to do a lot more on the client side with regards to data manipulation. For now the easiest thing to do is to continue replicating the catalyst functionality. we don't have too much longer till these endpoints are fixed and then we can start discussing other architectural things.